public/app/features/variables/query/variableQueryRunner.ts

```diff
@@ -0,0 +1,199 @@
+  getVariable: typeof getVariable;
+  getTemplatedRegex: typeof getTemplatedRegex;
+  getTimeSrv: typeof getTimeSrv;
+  queryRunners: QueryRunners;
+  runRequest: typeof runRequest;
+}
+
+export class VariableQueryRunner {
+  private readonly updateOptionsRequests: Subject<UpdateOptionsArgs>;
+  private readonly updateOptionsResults: Subject<UpdateOptionsResults>;
+  private readonly cancelRequests: Subject<{ identifier: VariableIdentifier }>;
+  private readonly subscription: Unsubscribable;
+
+  constructor(
+    private dependencies: VariableQueryRunnerArgs = {
+      dispatch,
+      getState,
+      getVariable,
+      getTemplatedRegex,
+      getTimeSrv,
+      queryRunners: new QueryRunners(),
+      runRequest,
+    }
+  ) {
+    this.updateOptionsRequests = new Subject<UpdateOptionsArgs>();
+    this.updateOptionsResults = new Subject<UpdateOptionsResults>();
+    this.cancelRequests = new Subject<{ identifier: VariableIdentifier }>();
+    this.onNewRequest = this.onNewRequest.bind(this);
+    this.subscription = this.updateOptionsRequests.subscribe(this.onNewRequest);
+  }
+
+  queueRequest(args: UpdateOptionsArgs): void {
+    this.updateOptionsRequests.next(args);
+  }
+
+  getResponse(identifier: VariableIdentifier): Observable<UpdateOptionsResults> {
+    return this.updateOptionsResults.asObservable().pipe(filter(result => result.identifier === identifier));
+  }
+
+  cancelRequest(identifier: VariableIdentifier): void {
+    this.cancelRequests.next({ identifier });
+  }
+
+  destroy(): void {
+    this.subscription.unsubscribe();
+  }
+
+  private onNewRequest(args: UpdateOptionsArgs): void {
+    const { datasource, identifier, searchFilter } = args;
+    try {
+      const {
+        dispatch,
+        runRequest,
+        getTemplatedRegex: getTemplatedRegexFunc,
+        getVariable,
+        queryRunners,
+        getTimeSrv,
+        getState,
+      } = this.dependencies;
+
+      const beforeUid = getState().templating.transaction.uid;
+
+      this.updateOptionsResults.next({ identifier, state: LoadingState.Loading });
+
+      const variable = getVariable<QueryVariableModel>(identifier.id, getState());
+      const timeSrv = getTimeSrv();
+      const runnerArgs = { variable, datasource, searchFilter, timeSrv, runRequest };
+      const runner = queryRunners.getRunnerForDatasource(datasource);
+      const target = runner.getTarget({ datasource, variable });
+      const request = this.getRequest(variable, args, target);
+
+      runner
+        .runRequest(runnerArgs, request)
+        .pipe(
+          filter(() => {
+            // lets check if we started another batch during the execution of the observable. If so we just want to abort the rest.
+            const afterUid = getState().templating.transaction.uid;
+            return beforeUid === afterUid;
+          }),
+          first(data => data.state === LoadingState.Done || data.state === LoadingState.Error),
+          toMetricFindValues(),
```

Comment:
If I understand the flow correctly- when the query results in an error state- we will still go through the entire pipe here, right? Is there a way to avoid that somehow and handle the error earlier?

Reply:
Good catch, I didn't bother with data.state === LoadingState.Error in this PR. Will look into that!

Reply:
Fixed!